examples/chat/src/network.rs

```diff
@@ -3,154 +3,114 @@ use std::time::Duration;
-                    _ => {
-                        error!("Recieved unexpected network event: {:?}", event)
-                    }
-                }
-            }
-            _ = tokio::time::sleep(Duration::from_secs(1)) => {
-                // Likely listening on all interfaces now, thus continuing by breaking the loop.
-                break;
-            }
-        }
+    for addr in swarm_listen {
+        client.listen_on(addr).await?;
     }
 
-    // Connect to the relay server. Not for the reservation or relayed connection, but to (a) learn
-    // our local public address and (b) enable a freshly started relay to learn its public address.
-    client.dial(relay_address.clone()).await?;
-
-    let mut learned_observed_addr = false;
-    let mut told_relay_observed_addr = false;
-    let relay_peer_id = match relay_address.iter().find_map(|protocol| {
-        if let libp2p::multiaddr::Protocol::P2p(peer_id) = protocol {
-            Some(peer_id)
-        } else {
-            None
-        }
-    }) {
-        Some(peer_id) => peer_id,
-        None => eyre::bail!("Failed to get PeerId from relay address"),
-    };
-
-    loop {
-        match event_receiver.recv().await.unwrap() {
-            types::NetworkEvent::IdentifySent { peer_id } => {
-                if peer_id == relay_peer_id {
-                    info!("Told relay its public address");
-                    told_relay_observed_addr = true;
-                }
-            }
-            types::NetworkEvent::IdentifyReceived {
-                peer_id,
-                observed_addr,
-            } => {
-                if peer_id == relay_peer_id {
-                    info!("Relay told us our observed address: {}", observed_addr);
-                    learned_observed_addr = true;
-                }
-            }
-            event => info!("unexpected: {event:?}"),
-        };
+    tokio::spawn(run_init_dial(
+        client.clone(),
+        rendezvous_addresses,
+        relay_addresses,
+    ));
 
-        if learned_observed_addr && told_relay_observed_addr {
-            break;
-        }
-    }
-
-    // Create reservation on relay server and wait for it to be accepted ...
-    client
-        .listen_on(relay_address.with(multiaddr::Protocol::P2pCircuit))
-        .await?;
+    Ok((client, event_receiver))
+}
 
-    loop {
-        match event_receiver.recv().await.unwrap() {
-            types::NetworkEvent::RelayReservationAccepted => {
-                info!("Relay accepted our reservation");
-                break;
-            }
-            event => info!("unexpected: {event:?}"),
+async fn run_init_dial(
+    client: NetworkClient,
+    rendezvous_addresses: Vec<Multiaddr>,
+    relay_addresses: Vec<Multiaddr>,
+) {
+    tokio::time::sleep(Duration::from_secs(5)).await;
+
+    for addr in rendezvous_addresses
+        .into_iter()
+        .chain(relay_addresses)
+        .collect::<std::collections::HashSet<_>>()
+        .into_iter()
```

Comment:
let's move this uniq requirement to the caller

either take a `HashSet` or dial what it's given and leave it to the caller to fulfil the uniq contract